docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
+
+Next, we'll add a circle:
+
+[!code-csharp[CircleDefinition](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#03_CircleDefinition "Circle definition")]
+
+We'll also write a method to compute the area of a circle:
+
+[!code-csharp[04_ComputeCircleArea](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#04_ComputeCircleArea "Compute circle area")]
+
+Ideally, you'd want one method that computes the area of any kind of
+shape. You've only created a couple different kinds of shapes so far.
+Let's use an `is` pattern matching expression to test the type of 
+shape:
+
+[!code-csharp[05_ComputeWithAsExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#05_ComputeWithAsExpression "Compute with as expression")]
+
+The parameter type for this method is `object`: any type can be passed
+to this method. The code in the method checks the type to see if the parameter
+is either a square or a circle. Notice that the pattern match `is` expression
+can assign a variable as well as test the type. The pattern match expression
+`shape is Square s` is semantically the same as the more verbose syntax:
+
+```csharp
+Square s;
+if (shape is Square)
+{
+    s = (shape as Square);
+    // use 's'
+} 
+```
+
+Language rules for pattern matching expressions help you avoid misuising
+the results of a match expression. In the example above, the variables `s`
+and `c` are only in scope and definitely assigned when the respective
+pattern match expressions have true results. If you try and use either
+variable in another location, your code generates compiler errors.
+
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
+`s` is in scope in the method `ComputeArea`. That's because the true and false
+branches of an `if` statement each establish a scope. However, the `if` statement
+itself does not. That means variables declared in the `if` statement are in the
+same scope as the `if` statement (the method in this case.) 
+
+The variables `c` and `s` are assigned when the respective `if` statements are true
+because of the definitely assigned when true mechanism. (You could reverse the
+logic by saying `if !(shape is Square s)` and the variable `s` would be definitely
+assigned only in the false branch.) 
+
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern ws not met.
+
+## Extending to more shape types
+
+As time goes on, you need to support other shape types. As the number
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions becomes cumbersome. In addition to requiring `if`
+statements on each type you want to check, the `is` expressions are limited
+to testing a type.
```

Comment:
That's an interesting philosophical question, I suppose.  If I have `if (x is BaseType b)` and pass in a type that derives from `BaseType`, though that expression matches for that entire type hierarchy, I'm still only testing against `BaseType`.  There's no way for me to observe `b` as anything other than a `BaseType`.

I think that given the above, it would make sense to say, "testing if the input matches a single type".